### ISSUE TITLE
[plan] Headless test suites leak ~1700 nodes at exit, masking any real leak a change introduces

### ISSUE THREAD

**Issue by stardustsuperwizard** [stardustsuperwizard/mikeys_game_bones-rules-moba#308]
## Summary

Nearly every suite under `rules/tests/` constructs `Actor`, `MobaCombatant`,
`MobaStateMachine` and similar `Node` instances standalone and never frees them. A
`Node` that is never added to a `SceneTree` is not reclaimed by the engine, so each
headless validation run ends with:

```
WARNING: 1715 ObjectDB instances were leaked at exit (run with `--verbose` for details).
ERROR: 43 resources still in use at exit (run with --verbose for details).
```

The leak is harmless to the suites themselves — they pass — but it destroys the signal.
A genuine leak introduced in production code would move the number from 1706 to, say,
1709 and nobody would notice. The baseline is the bug, not any individual instance of
it.

This was raised as a non-blocking finding by the automated reviewer on #307 and
deliberately left unfixed there, since #303 scoped that PR to the `owner_id` change and
the pattern predates it. Filing it here as the reviewer's finding, per
`CLAUDE.md` → *Working without an Issue*.

## Expected Behavior

A clean headless validation run ends with no `ObjectDB instances were leaked at exit`
warning and no `resources still in use at exit` error, so that either line appearing is
actionable evidence that a change leaked something.

## Actual Behavior

Every run — before and after any given change — emits both lines with a large baseline
count. Neither line fails the build: `.github/scripts/validate-godot.sh` gates on the
process exit code and on a `grep` for `All N test suites passed.`, and the leak is
reported by the engine at shutdown after the tests have already reported success. So the
run is green, the `ERROR:` line scrolls past in CI, and reviewers are trained to ignore
an error-level line in the validation output.

## Reproduction Steps

1. Check out `main` at `14a28af` (pre-#303).
2. `godot --headless --path . --import`
3. `godot --headless --path . --quit 2>&1 | grep -E "leaked at exit|resources still in use"`
4. Observe the leak warning and the resource error, with the run otherwise green.

## Reproduction Rate

Always. The count is deterministic across runs at a given commit.

## Environment

- Godot version: `4.7.1.stable.official.a13da4feb` (the version the workflows pin)
- Branch/commit: reproduced at `14a28af` (pre-#303) and `43987b6` (post-#303, merged as #307)
- Platform: Linux x86_64, headless
- Relevant configuration: `.github/scripts/validate-godot.sh`, `tests/test_bootstrap.gd`

## Evidence

Measured at two commits with the same binary:

| Commit | Suites | Leaked instances | Resources in use |
| --- | --- | --- | --- |
| `14a28af` (pre-#303) | 32 | 1706 | 43 |
| `43987b6` (post-#303) | 33 | 1715 | 43 |

The +9 is `rules/tests/authority_test.gd`, added by #307 — three tests × one `Actor` with
a `MobaCombatant` child, none freed. It is a representative new instance of the existing
pattern, not the cause.

The diagnosis was confirmed causally rather than inferred: adding an `actor.free()` to the
end of each of those three tests and re-running dropped the count from **1715 back to
exactly 1706**, with `All 33 test suites passed.` still reported. Freeing the root test
node reclaims its children, so the mechanical shape of a fix is one `free()` per
constructed root, not a rewrite of the suites.

Scale of the pattern — `.new()` calls on node types vs. `free()`/`queue_free()` calls, per
file:

| File | constructs | frees |
| --- | --- | --- |
| `rules/tests/combatant_test.gd` | 22 | 1 |
| `rules/tests/death_test.gd` | 20 | 10 |
| `rules/tests/cast_cancel_test.gd` | 18 | 2 |
| `rules/tests/crowd_control_test.gd` | 18 | 0 |
| `rules/tests/channel_test.gd` | 14 | 0 |
| `rules/tests/loadout_test.gd` | 14 | 0 |
| `rules/tests/cooldown_test.gd` | 13 | 0 |
| `rules/tests/resource_test.gd` | 10 | 0 |
| `rules/tests/ability_activation_test.gd` | 7 | 2 |
| `rules/tests/shield_bash_ability_test.gd` | 6 | 0 |
| `rules/tests/sustain_test.gd` | 4 | 0 |
| `rules/tests/authority_test.gd` | 2 | 0 |
| …and ~15 more | | |

Counts are a rough grep over node-type constructors and are indicative, not exact — some
suites (`state_machine_test.gd`, `hud_test.gd`, `target_frame_test.gd`) already free
consistently, which is where the convention should be read from.

Note that the 43 in-use resources are identical on both sides, so that line is entirely
pre-existing and may have a different root cause (likely `.tres` fixtures held by
`MobaAbilityLibrary`) than the node leak — worth confirming during planning rather than
assuming one fix addresses both.

## Suspected Scope (Optional)

`rules/tests/**` and `tests/test_bootstrap.gd`. Two questions worth settling during
planning, since they change the shape of the work considerably:

- Whether suites free their own fixtures, or whether `test_bootstrap.gd` gains a
  per-suite teardown seam so individual suites do not each have to remember.
- Whether the run should *fail* on a non-zero leak count once the baseline is clean. That
  is what converts this from a one-time cleanup into a standing guard, and it interacts
  with `validate-godot.sh`'s existing content check — the same gap #196 closed for
  truncated runs.

## Constraints

- Fix the underlying defect rather than suppressing the symptom.
- Do not weaken tests to make the issue disappear.
- Avoid unrelated refactoring.
- Suites must still pass unchanged — a fixture freed too early turns a passing assertion
  into a use-after-free, so freeing belongs after the last read of the fixture.
- `ability_activation_test.gd` deliberately frees a target mid-test to exercise
  `is_instance_valid()` guards; that existing `free()` is load-bearing and must not be
  disturbed.

## Acceptance Criteria

- [ ] Reproduction steps no longer reproduce the defect.
- [ ] Expected behavior is restored.
- [ ] Regression test added when practical.
- [ ] Existing tests pass.
- [ ] Godot headless validation passes.

## Out of Scope

- The 43 in-use resources at exit, if planning determines it has a separate root cause
  from the node leak — split it out rather than widening this.
- Any change to what the suites actually assert.
- Production-code lifetime changes; this is about test fixtures, not `Actor.die()` or
  `queue_free()` in game code (see `docs/planned_features.md` §0.3 for that separate
  latent bug).
